File: emm/web_ui/src/api/auth/passwordPolicy.ts
/**
 * Настраиваемая парольная политика логина (`auth_service`).
 *
 * Минимальная длина + обязательность буквы/цифры для пользовательских паролей.
 * Хранится singleton-строкой в БД, правит `account_admin`. На первом запуске
 * сервиса сидируется из env (`AUTH_PASSWORD_POLICY_*`), дальше — из БД.
 * `INITIAL_ADMIN_PASSWORD` этой политике не подчиняется (жёсткий guard = 12).
 *
 *   GET  /api/auth/v1/admin/password-policy  — текущая политика (account_admin)
 *   PUT  /api/auth/v1/admin/password-policy  — обновить (account_admin)
 *   GET  /api/auth/v1/password-policy         — публично (для форм ввода пароля)
 */

import { apiGet, apiPut } from "@/api/client";

/** Границы настраиваемой минимальной длины (совпадают с backend'ом). */
export const MIN_CONFIGURABLE_LENGTH = 1;
export const MAX_CONFIGURABLE_LENGTH = 128;

/** Текущая политика (ответ GET). */
export interface PasswordPolicy {
  min_length: number;
  require_letter: boolean;
  require_digit: boolean;
  updated_at?: string | null;
  updated_by?: string | null;
}

/** Тело PUT — частичное обновление. */
export type PasswordPolicyUpdate = Partial<
  Pick<PasswordPolicy, "min_length" | "require_letter" | "require_digit">
>;

const BASE = "/auth/v1";

/** Текущая политика логина (account_admin). 403 — нет роли. */
export function getPasswordPolicy(): Promise<PasswordPolicy> {
  return apiGet<PasswordPolicy>(`${BASE}/admin/password-policy`);
}

/** Обновить политику логина (частичная замена; account_admin). */
export function putPasswordPolicy(
  body: PasswordPolicyUpdate,
): Promise<PasswordPolicy> {
  return apiPut<PasswordPolicy>(`${BASE}/admin/password-policy`, body);
}

/** Публичная текущая политика — для форм ввода пароля (без авторизации). */
export function getPublicPasswordPolicy(): Promise<PasswordPolicy> {
  return apiGet<PasswordPolicy>(`${BASE}/password-policy`);
}

File: emm/server_service/src/schemas/password_policy.py
"""Pydantic-схемы настраиваемой парольной политики (`/admin/password-policy`).

Платформенный singleton под `account_admin`. Описывает базовую политику для
ручного ввода пароля server-аккаунтов и IPMI-credentials: минимальную длину и
обязательность буквы/цифры. Границы `min_length` enforce'ятся на уровне поля
(`ge`/`le`); PUT частичный — любое поле можно опустить, тогда текущее значение
сохраняется.
"""

from datetime import datetime

from pydantic import BaseModel, Field

from src.core.password_policy import (
    MAX_CONFIGURABLE_LENGTH,
    MIN_CONFIGURABLE_LENGTH,
)


class PasswordPolicyResponse(BaseModel):
    """Текущая базовая парольная политика."""

    min_length: int = Field(
        description="Минимальная длина пароля.",
    )
    require_letter: bool = Field(
        description="Требовать хотя бы одну букву.",
    )
    require_digit: bool = Field(
        description="Требовать хотя бы одну цифру.",
    )
    updated_at: datetime | None = Field(
        default=None,
        description="Когда политику последний раз меняли (UTC).",
    )
    updated_by: str | None = Field(
        default=None,
        description="Actor-id платформенного админа, применившего последнее изменение.",
    )


class PasswordPolicyUpdate(BaseModel):
    """Тело PUT — частичное обновление базовой парольной политики.

    Любое поле можно опустить — тогда текущее значение сохраняется. Границы
    `min_length` проверяются здесь; полностью «пустая» политика (короткая длина
    без требований к символам) допустима, это осознанный выбор админа.
    """

    min_length: int | None = Field(
        default=None,
        ge=MIN_CONFIGURABLE_LENGTH,
        le=MAX_CONFIGURABLE_LENGTH,
        description=(
            "Минимальная длина пароля. "
            f"Диапазон {MIN_CONFIGURABLE_LENGTH}..{MAX_CONFIGURABLE_LENGTH}."
        ),
    )
    require_letter: bool | None = Field(
        default=None,
        description="Требовать хотя бы одну букву.",
    )
    require_digit: bool | None = Field(
        default=None,
        description="Требовать хотя бы одну цифру.",
    )
